GitCommands/Git/GitGpgController.cs

```diff
@@ -0,0 +1,281 @@
+        string GetCommitVerificationMessage(GitRevision revision);
+
+        /// <summary>
+        /// Obtain the tag status on current revision.
+        /// </summary>
+        /// <returns>Enum value that indicate if current git revision has one tag with good signature, one tag with bad signature or more than one tag.</returns>
+        Task<TagStatus> GetRevisionTagSignatureStatusAsync(GitRevision revision);
+
+        /// <summary>
+        /// Obtain the tag verification message for all the tag in current git revision 
+        /// </summary>
+        /// <returns>Full concatenated string coming from GPG analysis on all tags on current git revision.</returns>
+        string GetTagVerifyMessage(GitRevision revision);
+    }
+
+
+    public class GitGpgController : IGitGpgController
+    {
+        private readonly Func<IGitModule> _getModule;
+
+        /* Commit GPG status */
+        private const string GoodSign = "G";
+        private const string BadSign = "B";
+        private const string UnkSignValidity = "U";
+        private const string ExpiredSign = "X";
+        private const string ExpiredSignKey = "Y";
+        private const string RevokedKey = "R";
+        private const string MissingPubKey = "E";
+        private const string NoSign = "N";
+
+
+        /* Tag GPG status */
+        private const string GoodSignature = "GOODSIG";
+        private const string ValidTagSign = "VALIDSIG";
+        private const string NoTagPubKey = "NO_PUBKEY";
+        private const string NoSignatureFound = "error: no signature found";
+
+        private static readonly Regex ValidSignatureTagRegex = new Regex(ValidTagSign, RegexOptions.Compiled);
+        private static readonly Regex GoodSignatureTagRegex = new Regex(GoodSignature, RegexOptions.Compiled);
+        private static readonly Regex NoPubKeyTagRegex = new Regex(NoTagPubKey, RegexOptions.Compiled);
+        private static readonly Regex NoSignatureFoundTagRegex = new Regex(NoSignatureFound, RegexOptions.Compiled);
+
+        /// <summary>
+        /// Obtain the tag verification message for all the tag in current git revision 
+        /// </summary>
+        /// <returns>Full concatenated string coming from GPG analysis on all tags on current git revision.</returns>
+        public GitGpgController(Func<IGitModule> getModule)
+        {
+            _getModule = getModule;
+        }
+
+
+        /// <summary>
+        /// Obtain the commit signature status on current revision.
+        /// </summary>
+        /// <returns>Enum value that indicate the gpg status for current git revision.</returns>
+        public async Task<CommitStatus> GetRevisionCommitSignatureStatusAsync(GitRevision revision)
+        {
+            if (revision == null)
+            {
+                throw new ArgumentNullException("revision");
+            }
+
+            var module = GetModule();
+            return await Task.Run(() =>
```

Comment:
This is something I've missed and @jbialobr picked up - GitModule is not thread-safe, so we need to make a copy of it before passing to a background thread.
Similar to https://github.com/gitextensions/gitextensions/blob/c133bf68944be5e68fcdc66dc7dcdc4c792c222e/GitUI/CommandsDialogs/FormBrowse.cs#L2417-L2422

Reply:
Changed.

Reply:
This change causes `Validate_GetRevisionCommitSignatureStatusAsync` tests to fail.

Because the tests aren't written correctly, you don't see them fail.
We use nunit 3.8, and instead of `async void` we must use `async Task` decorations. For more info please see https://www.thomaslevesque.com/2015/02/01/async-unit-tests-with-nunit/

Once the tests are changed, e.g.:

```
        [TestCase(CommitStatus.GoodSignature, "G")]
        [TestCase(CommitStatus.SignatureError, "B")]
        [TestCase(CommitStatus.SignatureError, "U")]
        [TestCase(CommitStatus.SignatureError, "X")]
        [TestCase(CommitStatus.SignatureError, "Y")]
        [TestCase(CommitStatus.SignatureError, "R")]
        [TestCase(CommitStatus.MissingPublicKey, "E")]
        [TestCase(CommitStatus.NoSignature, "N")]
        public async Task Validate_GetRevisionCommitSignatureStatusAsync(CommitStatus expected, string gitCmdReturn)
        {
            var guid = Guid.NewGuid().ToString("N");
            GitRevision revision = new GitRevision(null, guid);

            _module().RunGitCmd($"log --pretty=\"format:%G?\" -1 {revision.Guid}").Returns(gitCmdReturn);

            var actual = await _gpgController.GetRevisionCommitSignatureStatusAsync(revision);
            Assert.AreEqual(expected, actual);
        }
```

you see them correctly capture the result:
![image](https://user-images.githubusercontent.com/4403806/33522262-82190818-d83c-11e7-97f1-e0e1fadb08be.png)

. . .

The underlying reason for the failure is the change I asked you to do (GitGpgController.cs, line 109). This makes it non mockable, so the arrangements set for `module.RunGitCmd` don't apply....
This is a current design issue, which I'm trying to work through...

For your PR, let's roll the last change back to the original:

```
            var module = GetModule();
            return await Task.Run(() =>
            {
                CommitStatus cmtStatus;

                string gpg = module.RunGitCmd($"log --pretty=\"format:%G?\" -1 {revision.Guid}");

```

Sorry about this.